src/CowAmmModule.sol

```diff
@@ -0,0 +1,299 @@
+     */
+    function replaceAmm(
+        IERC20 token0,
+        IERC20 token1,
+        uint256 minTradedToken0,
+        address priceOracle,
+        bytes calldata priceOracleData,
+        bytes32 appData
+    ) external {
+        // Assume the caller is a Safe
+        Safe safe = Safe(payable(msg.sender));
+
+        bytes32 _activeOrder = activeOrders[safe];
+        if (_activeOrder != EMPTY_AMM_HASH) {
+            _closeAmm(safe, _activeOrder);
+        }
+        _createAmm(safe, token0, token1, minTradedToken0, priceOracle, priceOracleData, appData);
+    }
+
+    /**
+     * @notice Closes the active CoW AMM.
+     * @dev This function will call `ComposableCoW.remove` with the active order if there is one.
+     */
+    function closeAmm() external {
+        Safe safe = Safe(payable(msg.sender));
+        bytes32 _activeOrder = activeOrders[safe];
+        if (_activeOrder != EMPTY_AMM_HASH) {
+            _closeAmm(safe, _activeOrder);
+        }
+    }
+
+    // --- Internal functions
+
+    /**
+     * @notice Creates a new CoW AMM with the given parameters.
+     * @param safe The address of the Safe that owns the new CoW AMM.
+     * @param token0 The address of the first token in the pair.
+     * @param token1 The address of the second token in the pair.
+     * @param minTradedToken0 The minimum amount of token0 before the AMM attempts auto-rebalance.
+     * @param priceOracle The address of the price oracle to use for the AMM.
+     * @param priceOracleData The data to pass to the price oracle.
+     * @param appData The app data to pass to the AMM.
+     */
+    function _createAmm(
+        Safe safe,
+        IERC20 token0,
+        IERC20 token1,
+        uint256 minTradedToken0,
+        address priceOracle,
+        bytes calldata priceOracleData,
+        bytes32 appData
+    ) internal {
+        // Always make sure the module is setup before doing anything
+        _setup(safe);
+
+        // We can't create a new CoW AMM if there isn't anything to trade!
+        if (token0.balanceOf(address(safe)) == 0 || token1.balanceOf(address(safe)) == 0) {
+            revert TokenBalanceZero();
+        }
+
+        // Set the token allowances for the vault relayer
+        _setAllowance(safe, token0);
+        _setAllowance(safe, token1);
+
+        // Prepare the data for the new CoW AMM
+        ConstantProduct.Data memory data = ConstantProduct.Data({
+            token0: token0,
+            token1: token1,
+            minTradedToken0: minTradedToken0,
+            priceOracle: IPriceOracle(priceOracle),
+            priceOracleData: priceOracleData,
+            appData: appData
+        });
+
+        // Wrap the data in a conditional order
+        IConditionalOrder.ConditionalOrderParams memory params = IConditionalOrder.ConditionalOrderParams({
+            handler: HANDLER,
+            salt: keccak256(abi.encodePacked(block.timestamp)),
```

Comment:
nit: keccka data as well to allow multiple deployments of different pairs in the same block?

Reply:
I don't think this is relevant, as we don't want to allow a Safe to have more than 1 CoW AMM running at a time (therefore, no need for deployments of different pairs)?